pkg/cloud/aws/actuators/cluster/actuator.go

```diff
@@ -64,14 +64,48 @@ func (a *Actuator) Reconcile(cluster *clusterv1.Cluster) error {
 	elbsvc := elb.NewService(scope)
 
 	// Store some config parameters in the status.
-	if len(scope.ClusterConfig.CACertificate) == 0 {
-		caCert, caKey, err := certificates.NewCertificateAuthority()
+	if !scope.ClusterConfig.CAKeyPair.HasCertAndKey() {
+		caCert, caKey, err := actuators.GetOrGenerateKeyPair(&scope.ClusterConfig.CAKeyPair, actuators.ClusterCA)
 		if err != nil {
 			return errors.Wrap(err, "Failed to generate a CA for the control plane")
 		}
+		scope.ClusterConfig.CAKeyPair = v1alpha1.KeyPair{
+			Cert: caCert,
+			Key:  caKey,
+		}
+	}
+
+	if !scope.ClusterConfig.EtcdCAKeyPair.HasCertAndKey() {
+		etcdCACert, etcdCAKey, err := actuators.GetOrGenerateKeyPair(&scope.ClusterConfig.EtcdCAKeyPair, actuators.EtcdCA)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get or generate etcd CA cert")
+		}
+		scope.ClusterConfig.EtcdCAKeyPair = v1alpha1.KeyPair{
+			Cert: etcdCACert,
+			Key:  etcdCAKey,
+		}
+	}
 
-		scope.ClusterConfig.CACertificate = certificates.EncodeCertPEM(caCert)
-		scope.ClusterConfig.CAPrivateKey = certificates.EncodePrivateKeyPEM(caKey)
+	if !scope.ClusterConfig.FrontProxyCAKeyPair.HasCertAndKey() {
+		fpCACert, fpCAKey, err := actuators.GetOrGenerateKeyPair(&scope.ClusterConfig.FrontProxyCAKeyPair, actuators.FrontProxyCA)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get or generate front-proxy CA cert")
+		}
+		scope.ClusterConfig.FrontProxyCAKeyPair = v1alpha1.KeyPair{
+			Cert: fpCACert,
+			Key:  fpCAKey,
+		}
+	}
+
+	if !scope.ClusterConfig.SAKeyPair.HasCertAndKey() {
+		saPub, saKey, err := actuators.GetOrGenerateKeyPair(&scope.ClusterConfig.SAKeyPair, actuators.ServiceAccount)
```

Comment:
It seems a bit odd here to handle the SA keypair similarly to the CA certificate keypairs. I think we should probably special case the handling of the SA keypair since the generation, file extensions, and usages are different.

I'm good with tracking that as a followup item, though.

Reply:
Yes, I also want to suggest that we use k8s.io/client-go/util/cert instead of writing our own.